context.go

```diff
@@ -0,0 +1,239 @@
+
+	c.ResponseWriter.WriteHeader(status)
+}
+
+func (c *hcontext) Write(data []byte) (int, error) {
+	if c.closed {
+		panic(fmt.Errorf("Trying to write to response after WriteModel or WriteError for %s %s", c.r.Method, c.r.URL.Path))
+	}
+
+	return c.ResponseWriter.Write(data)
+}
+
+func (c *hcontext) WriteError(status int, message string, errors ...error) {
+	if c.closed {
+		panic(fmt.Errorf("Trying to write to response after WriteModel or WriteError for %s %s", c.r.Method, c.r.URL.Path))
+	}
+
+	details := []*ErrorDetail{}
+
+	c.errors = append(c.errors, errors...)
+	for _, err := range c.errors {
+		if d, ok := err.(ErrorDetailer); ok {
+			details = append(details, d.ErrorDetail())
+		} else {
+			details = append(details, &ErrorDetail{Message: err.Error()})
+		}
+	}
+
+	model := &ErrorModel{
+		Title:  http.StatusText(status),
+		Status: status,
+		Detail: message,
+		Errors: details,
+	}
+
+	// Select content type and transform it to the appropriate error type.
+	ct := selectContentType(c.r)
+	switch ct {
+	case "application/cbor":
+		ct = "application/problem+cbor"
+	case "", "application/json":
+		ct = "application/problem+json"
+	case "application/yaml", "application/x-yaml":
+		ct = "application/problem+yaml"
+	}
+
+	c.writeModel(ct, status, model)
+}
+
+func (c *hcontext) WriteModel(status int, model interface{}) {
+	if c.closed {
+		panic(fmt.Errorf("Trying to write to response after WriteModel or WriteError for %s %s", c.r.Method, c.r.URL.Path))
+	}
+
+	// Get the negotiated content type the client wants and we are willing to
+	// provide.
+	ct := selectContentType(c.r)
+
+	c.writeModel(ct, status, model)
+}
+
+func (c *hcontext) writeModel(ct string, status int, model interface{}) {
+	// Is this allowed? Find the right response.
+	if c.op != nil {
+		responses := []Response{}
+		names := []string{}
+		statuses := []string{}
+		for _, r := range c.op.responses {
+			statuses = append(statuses, fmt.Sprintf("%d", r.status))
+			if r.status == status {
+				responses = append(responses, r)
+				if r.model != nil {
+					names = append(names, r.model.Name())
+				}
+			}
+		}
+
+		if len(responses) == 0 {
+			panic(fmt.Errorf("HTTP status %d not allowed for %s %s, expected one of %s", status, c.r.Method, c.r.URL.Path, statuses))
+		}
+
+		found := false
+		for _, r := range responses {
+			if r.model == reflect.TypeOf(model) {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			panic(fmt.Errorf("Invalid model %s, expecting %s for %s %s", reflect.TypeOf(model), strings.Join(names, ", "), c.r.Method, c.r.URL.Path))
+		}
+	}
+
+	// Do the appropriate encoding.
+	var encoded []byte
+	var err error
+	if strings.HasPrefix(ct, "application/json") || strings.HasSuffix(ct, "+json") {
+		encoded, err = json.Marshal(model)
+		if err != nil {
+			panic(fmt.Errorf("Unable to marshal JSON: %w", err))
+		}
+	} else if strings.HasPrefix(ct, "application/yaml") || strings.HasPrefix(ct, "application/x-yaml") || strings.HasSuffix(ct, "+yaml") {
```

Comment:
[RFC 6839 Section 3.1](https://tools.ietf.org/html/rfc6839#section-3.1) says that:

>[RFC4627] defines the "application/json" media type.  The suffix "+json" MAY be used with any media type whose representation follows that established for "application/json".

The same is true for the `+yaml` suffix. I guess that means `application/json+yaml` might be valid, but in practice nobody uses it and it *should* be parsed as YAML.

Reply:
Gotcha; I wasn't sure if there was the possibility of needing to support multiple return types and the way it was written made it seem like it could be. Good to know it's not!